File: lib/ftw_realtime/marketplace/job.ex
defmodule FtwRealtime.Marketplace.Job do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id

  schema "jobs" do
    field(:title, :string)
    field(:description, :string)
    field(:category, :string)
    field(:budget_min, :integer)
    field(:budget_max, :integer)
    field(:location, :string)

    field(:status, Ecto.Enum,
      values: [:open, :bidding, :awarded, :in_progress, :completed, :disputed, :cancelled],
      default: :open
    )

    field(:bid_count, :integer, default: 0)

    belongs_to(:homeowner, FtwRealtime.Marketplace.User)
    has_many(:bids, FtwRealtime.Marketplace.Bid)

    timestamps(type: :utc_datetime)
  end

  def changeset(job, attrs) do
    job
    |> cast(attrs, [
      :title,
      :description,
      :category,
      :budget_min,
      :budget_max,
      :location,
      :status,
      :homeowner_id
    ])
    |> validate_required([:title, :description, :category, :location, :homeowner_id])
    |> validate_number(:budget_min, greater_than_or_equal_to: 0)
    |> validate_number(:budget_max, greater_than_or_equal_to: 0)
    |> foreign_key_constraint(:homeowner_id)
  end
end

File: src/main/kotlin/com/strata/ftw/domain/entity/Job.kt
package com.strata.ftw.domain.entity

import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.Instant
import java.util.UUID

@Entity
@Table(name = "jobs")
class Job(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "uuid")
    var id: UUID? = null,

    @Column(nullable = false)
    var title: String = "",

    var description: String? = null,

    var category: String? = null,

    @Column(name = "budget_min")
    var budgetMin: Int? = null,

    @Column(name = "budget_max")
    var budgetMax: Int? = null,

    var location: String? = null,

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    var status: JobStatus = JobStatus.open,

    @Column(name = "bid_count")
    var bidCount: Int = 0,

    var latitude: Double? = null,
    var longitude: Double? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "homeowner_id")
    var homeowner: User? = null,

    @Column(name = "homeowner_id", insertable = false, updatable = false)
    var homeownerId: UUID? = null,

    @OneToMany(mappedBy = "job", fetch = FetchType.LAZY)
    var bids: MutableList<Bid> = mutableListOf(),

    @CreationTimestamp
    @Column(name = "inserted_at", updatable = false)
    var insertedAt: Instant? = null,

    @UpdateTimestamp
    @Column(name = "updated_at")
    var updatedAt: Instant? = null
)

enum class JobStatus {
    open, bidding, awarded, in_progress, completed, disputed, cancelled
}
